### ISSUE TITLE
Softmax Output or not

### ISSUE THREAD

**Issue by michaelbzms** [slp-ntua/patrec-labs#72]
Στο https://twitter.com/karpathy/status/1013244313327681536 από τα references λέει ότι πιθανό λάθος είναι το "4) you passed softmaxed outputs to a loss that expects raw logits.".  Γιατί κάτι τέτοιο είναι λάθος; Δύο φορές το softmax δεν θα δόσει το ίδιο αποτέλεσμα;

Το CrossEntropyLoss, από ότι είχαμε βρει και στην προηγούμενη άσκηση κάνει και το ίδιο Softmax. Όμως μου φαίνεται πολύ περιέργο να αφήσω τo output raw για να γίνει το softmax στο loss αφού πχ στο predict, όταν δεν εχουμε loss calculation, δεν θα είχαμε πιθανότητες χωρίς Softmax. Άρα 1) Γιατί το nn.CrossEntropyLoss κάνει Softmax χωρις καν να είναι υπό επιλογή? 2) Είναι πρόβλημα να κάνουμε κι εμείς Softmax?

**Comment by filby89** [comment]
Καλησπέρα,
δυο φορές το softmax δεν δίνει το ίδιο αποτέλεσμα:
Έστω ότι έχουμε 3 κλάσεις ένα sample που ανήκει στην δεύτερη κλάση και έναν ταξινομητή. Αν ο ταξινομητής βγάλει σκορ (3.5, 6, 4) και κάνουμε softmax παίρνουμε (0.07, 0.82, 0.11), και αν ξανάκανουμε: (0.23,0.51, 0.25). Αυτό οδηγεί σε λανθασμένα σκορ (έρχονται πιο κοντά) και επίσης πιο αργό training. 

Το CrossEntropyLoss περιλαμβάνει το softmax για λόγους conveniency, όμως υπάρχει και το σκέτο NLLLoss (crossentropyloss = softmax+nllloss) αν το προτιμάτε.

Στο predicttion σας ενδιαφέρει μόνο το argmax το οποίο είναι ίδιο είτε κάνετε softmax είτε όχι.
